### DISCUSSION TITLE
Problem with Strava provider

### DISCUSSION THREAD

**Discussion by johann-arias** [nextauthjs/next-auth#5279] [Help]
### Question 💬

Hello guys, 

I'm new to next.js and next-auth and I'm trying to implet a sign-in with Strava feature. I followed this tutorial for Github : https://vercel.com/guides/nextjs-prisma-postgres and tried to make it work with Strava

You can find my code here : https://github.com/johann-arias/strava-stats

When I click the "Sign in with Strava" button, I get an error displayed in the browser : 

![Capture d’écran 2022-09-01 à 15 53 14](https://user-images.githubusercontent.com/25004937/187931515-6aa54144-586a-437e-9b3b-98293b7e58e3.png)

And here's what's my terminals shows : 

```
 Error:
Invalid `p.account.create()` invocation in
/Users/johann/Documents/dev/blogr-nextjs-prisma/node_modules/@next-auth/prisma-adapter/dist/index.js:19:42

  16 },
  17 updateUser: ({ id, ...data }) => p.user.update({ where: { id }, data }),
  18 deleteUser: (id) => p.user.delete({ where: { id } }),
→ 19 linkAccount: (data) => p.account.create({
       data: {
         provider: 'strava',
         type: 'oauth',
         providerAccountId: '49310442',
         token_type: 'Bearer',
         expires_at: 1662057083,
         refresh_token: '4cedaeac0b879c9938feb29aedf864948fe8c7b6',
         access_token: '4e176a36093404b850eedc4f0b6896f89d4cc309',
         athlete: {
         ~~~~~~~
           id: 49310442,
           username: null,
           resource_state: 2,
           firstname: 'Johann',
           lastname: 'Arias',
           bio: null,
           city: 'Paris',
           state: 'Île-de-France',
           country: 'France',
           sex: null,
           premium: true,
           summit: true,
           created_at: '2019-12-28T13:51:16Z',
           updated_at: '2022-08-19T16:00:14Z',
           badge_type_id: 1,
           weight: null,
           profile_medium: 'https://dgalywyr863hv.cloudfront.net/pictures/athletes/49310442/13342948/1/medium.jpg',
           profile: 'https://dgalywyr863hv.cloudfront.net/pictures/athletes/49310442/13342948/1/large.jpg',
           friend: null,
           follower: null
         },
         userId: 'cl7j2tl2a0026d7whtttyver6'
       }
     })

Unknown arg `athlete` in data.athlete for type AccountUncheckedCreateInput. Did you mean `type`? Available args:
type AccountUncheckedCreateInput {
  id?: String
  userId: String
  type: String
  provider: String
  providerAccountId: String
  refresh_token?: String | Null
  access_token?: String | Null
  expires_at?: Int | Null
  token_type?: String | Null
  scope?: String | Null
  id_token?: String | Null
  session_state?: String | Null
  oauth_token_secret?: String | Null
  oauth_token?: String | Null
}


    at Document.validate (/Users/johann/Documents/dev/blogr-nextjs-prisma/node_modules/@prisma/client/runtime/index.js:28174:20)
    at serializationFn (/Users/johann/Documents/dev/blogr-nextjs-prisma/node_modules/@prisma/client/runtime/index.js:30774:19)
    at runInChildSpan (/Users/johann/Documents/dev/blogr-nextjs-prisma/node_modules/@prisma/client/runtime/index.js:24002:12)
    at PrismaClient._executeRequest (/Users/johann/Documents/dev/blogr-nextjs-prisma/node_modules/@prisma/client/runtime/index.js:30781:31)
    at consumer (/Users/johann/Documents/dev/blogr-nextjs-prisma/node_modules/@prisma/client/runtime/index.js:30707:23)
    at /Users/johann/Documents/dev/blogr-nextjs-prisma/node_modules/@prisma/client/runtime/index.js:30712:51
    at AsyncResource.runInAsyncScope (node:async_hooks:203:9)
    at /Users/johann/Documents/dev/blogr-nextjs-prisma/node_modules/@prisma/client/runtime/index.js:30712:29
    at runInChildSpan (/Users/johann/Documents/dev/blogr-nextjs-prisma/node_modules/@prisma/client/runtime/index.js:24002:12)
    at PrismaClient._request (/Users/johann/Documents/dev/blogr-nextjs-prisma/node_modules/@prisma/client/runtime/index.js:30709:22) {
  name: 'LinkAccountError',
  code: undefined
}
```

A user is created in my database : 

![Capture d’écran 2022-09-01 à 15 56 03](https://user-images.githubusercontent.com/25004937/187932023-fe05e04b-6ff4-413e-89b1-9385e5af2057.png)

Thanks a lot for your help

### How to reproduce ☕️

https://github.com/johann-arias/strava-stats

go to localhost:3000, click on "Log In" and "Sign in with Strava"

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by ThangHuuVu** [comment]
It seems like you need to add athlete to your schema. 🤔

**Comment by schweden1997** [comment]
I ran into the same problem yesterday. Strava responds to a token request with the following  object.
```
{
  "token_type": "Bearer",
  "expires_in": 21600,
  "refresh_token": "e5n567567...",
  "access_token": "a4b945687g...",
  "athlete": {
    #{summary athlete representation}
  }
}
```
This object, including the `athlete` object, is spread into the `account` object in the `getProfile()` function. When Prisma tries to create the database entries from this `account` object it fails since `athlete` is not in the schema.

https://github.com/nextauthjs/next-auth/blob/92b38ed74003295c46544fbd62c3d22890f3bf0f/packages/next-auth/src/core/lib/oauth/callback.ts#L187-L196


Like @ThangHuuVu  suggested, this could be fixed by adding `athlete Json` to the `Account` model. (Edit: this might not work if the DB doesnt support the JSON type)

But since I don't actually want to store this information I used a custom `token` option in the `StravaProvider()`. This removes the `athlete` object from the `tokens` object.
```
{
  token: {
    async request({ client, params, checks, provider }) {
      const { token_type, expires_at, refresh_token, access_token } = await client.oauthCallback(provider.callbackUrl, params, checks)
      return {
        tokens: { token_type, expires_at, refresh_token, access_token },
      }
    },
  },
}
```

A suggestion would be to expand on the documentation or use this `token` option to remove the `athlete` object. 
I am happy to help but don't know where to start as I have never contributed to an OSS project.

**Reply by ThangHuuVu** [reply]
@schweden1997 sorry for the delayed response - this will be fixed after https://github.com/nextauthjs/next-auth/pull/5461 is merged. Essentially, we will be narrowing down what part of the account should be saved to the database.

**Reply by bursteways** [reply]
I'm still getting an error with the proposed solution above:

```
    StravaProvider({
      clientId: process.env.STRAVA_CLIENT_ID!,
      clientSecret: process.env.STRAVA_CLIENT_SECRET!,
      token: {
        async request({ client, params, checks, provider }) {
          const { token_type, expires_at, refresh_token, access_token } =
            await client.oauthCallback(provider.callbackUrl, params, checks);
          return {
            tokens: { token_type, expires_at, refresh_token, access_token },
          };
        },
      },
    }),
```

As it stands right now, Strava Provider can't be used unless we add the large `athlete` to our schema.

Looks like that PR was closed @ThangHuuVu - I checked the docs but didn't see a way to omit pieces of the oauth callback. Is that still being worked on?

**Reply by schweden1997** [reply]
I just tried my solution with a new bootstrapped t3 app and it still works. Can you provide some error logs.

**Comment by jhackett1** [comment]
still an issue as of april 2023. looks like pr #5461 didn't go anywhere

**Comment by ow1337** [comment]
Still same issue?

I get also sign in with a another user.. 
My error message is:
`Unknown argument `id`. Available options are listed in green. PrismaClientValidationError:
Invalid `prisma.account.create()` invocation:

{
  data: {
    provider: "strava",
    type: "oauth",
    providerAccountId: "1231234",
    token_type: "Bearer",
    expires_at: 1690328864,
    refresh_token: "123123d123123123d",
    access_token: "123123123123213ddee123123",
    athlete: {
      id: 1114444,
      ~~`

My schema.prisma:
`// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

model User {
  id             String    @id @default(auto()) @map("_id") @db.ObjectId
  username       String?   @unique
  email          String?   @unique
  name           String?
  emailVerified  DateTime?
  image          String?
  password       String?
  createdAt      DateTime  @default(now())
  updatedAt      DateTime  @updatedAt

  accounts Account[]
}

model Account {
  id                String  @id @default(auto()) @map("_id") @db.ObjectId
  userId            String  @db.ObjectId
  type              String
  provider          String
  providerAccountId String
  refresh_token     String? @db.String
  access_token      String? @db.String
  expires_at        Int?
  token_type        String?
  scope             String?
  id_token          String? @db.String
  session_state     String?

  athlete           Athlete? // Hier wurde das athlete Feld hinzugefügt

  user User @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@unique([provider, providerAccountId])
}

model Athlete {
  id               String     @id @default(auto()) @map("_id") @db.ObjectId
  username         String?
  resource_state   Int?
  firstname        String?
  lastname         String?
  bio              String?
  city             String?
  state            String?
  country          String?
  sex              String?
  premium          Boolean?
  summit           Boolean?
  created_at       String?
  updated_at       String?
  badge_type_id    Int?
  weight           String?
  profile_medium   String?
  profile          String?
  friend           String?
  follower         String?

  account Account? @relation(fields: [accountAthleteId], references: [id])
  accountAthleteId String? @unique @db.ObjectId
}
`

**Comment by madsenmm** [comment]
This is still an issue, any updates on this?
